dvc/external_repo.py

```diff
@@ -1,50 +1,111 @@
 from __future__ import unicode_literals
 
 import os
-import logging
 import tempfile
+from distutils.dir_util import copy_tree
 
 from funcy import retry
 from contextlib import contextmanager
 
-from dvc.config import Config
-from dvc.cache import CacheConfig
 from dvc.utils import remove
 
 
-logger = logging.getLogger(__name__)
+REPO_CACHE = {}
+REPO_BY_URL = {}
 
 
-def _clone(cache_dir=None, url=None, rev=None, rev_lock=None):
+@contextmanager
+def external_repo(url=None, rev=None, rev_lock=None, cache_dir=None):
     from dvc.repo import Repo
 
-    _path = tempfile.mkdtemp("dvc-repo")
+    path = _external_repo(url=url, rev=rev_lock or rev, cache_dir=cache_dir)
+    repo = Repo(path)
+    yield repo
+    repo.close()
 
-    repo = Repo.clone(url, _path, rev=(rev_lock or rev))
 
-    try:
-        if cache_dir:
-            cache_config = CacheConfig(repo.config)
-            cache_config.set_dir(cache_dir, level=Config.LEVEL_LOCAL)
-    finally:
-        repo.close()
+def _external_repo(url=None, rev=None, cache_dir=None):
+    key = (url, rev, cache_dir)
+    if key in REPO_CACHE:
+        return REPO_CACHE[key]
 
-    return Repo(_path)
+    new_path = tempfile.mkdtemp("dvc-erepo")
 
+    # Copy and adjust existing clone
+    if url in REPO_BY_URL:
+        old_path, old_rev, old_cache_dir = REPO_BY_URL[url]
 
-def _remove(repo):
-    repo.close()
+        # This one unlike shutil.copytree() works with an existing dir
+        copy_tree(old_path, new_path)
```

Comment:
This seems pretty weird. You are reinitilizing repo like 3 times. Why can't you just `Repo.clone(old_path, new_path, rev=rev)` or something like that? Seems like that would unify a lot of logic.

Reply:
Ah, right. I meant cloning from the local path instead of copying it manually. So not `old_path`, but mmm `REPO_CACHE[REPO_BY_URL[url]]`?

Reply:
`REPO_CACHE[REPO_BY_URL[url]]` is the same as `REPO_BY_URL[url][0]`

Reply:
And cloning from local dir is not the same as copying it, we will loose any non commited data, like local config.

Reply:
@Suor but we are editing local config anyway to set another cache_dir or reset it. We won't need to reset it if we clone.

Reply:
Copy doesn't need to know about all those implementation details, clone needs, so using  clone here will make code more fragile.